lib/spack/spack/solver/concretize.lp

```diff
@@ -461,13 +461,58 @@ error(100, "Attempted to use external for '{0}' which does not satisfy any confi
 % Config required semantics
 %-----------------------------------------------------------------------------
 
-activate_requirement_rules(Package) :- attr("node", Package).
-activate_requirement_rules(Package) :- attr("virtual_node", Package).
+activate_requirement(Package, X) :-
+  attr("node", Package),
+  requirement_group(Package, X),
+  not requirement_conditional(Package, X).
+
+activate_requirement(Package, X) :-
+  attr("virtual_node", Package),
+  requirement_group(Package, X),
+  not requirement_conditional(Package, X).
```

Comment:
It would be nice if we could derive a single fact from `node` or `virtual` node to reduce duplication:

```
attr("build", Package) :- attr("node", Package)
attr("build", Package) :- attr("virtual_node", Package)

activate_requirement(Package, X) :-
   attr("build", Package),
   requirement_group(Package, X),
   not requirement_conditional(Package, X).
```

Reply:
Done in 5a9898d0b4fc8aa6c5265a06c8a56265327facd2